lib/plug/body_reader/deflate.ex

```diff
@@ -0,0 +1,52 @@
+defmodule Plug.BodyReader.Deflate do
+  def init(%Plug.Conn{} = conn, _opts) do
+    zlib_stream =
+      case content_encoding(conn) do
+        encoding when encoding in [:zlib, :gzip] ->
+          zlib_stream = :zlib.open()
+          :ok = :zlib.inflateInit(zlib_stream, window_bits_for_encoding(encoding))
+          zlib_stream
+
+        :none ->
+          nil
+      end
+
+    private =
+      conn.private
+      |> Map.put(__MODULE__, zlib_stream)
+
+    {:ok, %Plug.Conn{conn | private: private}}
+  end
+
+  def close(%Plug.Conn{private: %{__MODULE__ => zlib_stream}} = conn, _opts) do
+    if !is_nil(zlib_stream) do
+      :zlib.close(zlib_stream)
+    end
+
+    {:ok, %Plug.Conn{conn | private: Map.delete(conn.private, __MODULE__)}}
+  end
+
+  def read_body(%Plug.Conn{private: %{__MODULE__ => nil}} = conn, opts) do
+    Plug.Conn.read_body(conn, opts)
+  end
+
+  def read_body(%Plug.Conn{private: %{__MODULE__ => zlib_stream}} = conn, opts) do
+    with {:ok, body, conn} <- Plug.Conn.read_body(conn, opts) do
+      case :zlib.safeInflate(zlib_stream, body) do
+        {_, data} -> {:ok, IO.iodata_to_binary(data), conn}
```

Comment:
A potential issue arises with this line if we read from the socket only a few bytes at a time because `:zlib.safeInflate/2` may then return `{:finished, []}` if more data is required before it can inflate it (I am not sure why it returns `:finished` instead of `:continue` but this is what I observed testing manually in the console).

It would be "surprising" to return `""` despite having actually read from the socket and in fact `Plug.Conn.next_multipart/2` would then erroneously raise an error.

The alternative would be to match on `[]` and recursively try to read more from the socket but this isn't ideal either as it breaks the expectation of `read_body` doing only one "read".